bundles/org.openhab.binding.mpd/src/main/java/org/openhab/binding/mpd/internal/protocol/MPDConnectionThread.java

```diff
@@ -0,0 +1,283 @@
+                    updateThingStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                } catch (MPDException e) {
+                    updateThingStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+                }
+
+                isInIdle.set(false);
+                closeSocket();
+
+                if (!disposed.get()) {
+                    sleep(RECONNECTION_TIMEOUT);
+                }
+            }
+        } catch (InterruptedException e) {
+        }
+    }
+
+    /**
+     * dispose the connection
+     */
+    public void dispose() {
+        disposed.set(true);
+        synchronized (pendingCommands) {
+            pendingCommands.add(new MPDCommand("close"));
+            sendNoIdleIfInIdle();
+            interrupt();
+        }
+    }
+
+    /**
+     * add a command to the pending commands queue
+     *
+     * @param command command to add
+     */
+    public void addCommand(MPDCommand command) {
+        insertCommand(command, -1);
+    }
+
+    private void insertCommand(MPDCommand command, int position) {
+        logger.debug("insert command '{}' at position {}", command.getCommand(), position);
+        int index = position;
+        synchronized (pendingCommands) {
+            if (index < 0) {
+                index = pendingCommands.size();
+            }
+            pendingCommands.add(index, command);
+            sendNoIdleIfInIdle();
+        }
+    }
+
+    private void updateThingStatus(ThingStatus status, ThingStatusDetail statusDetail, @Nullable String description) {
+        if (!disposed.get()) {
+            listener.updateThingStatus(status, statusDetail, description);
+        }
+    }
+
+    private void sendNoIdleIfInIdle() {
+        if (isInIdle.compareAndSet(true, false)) {
+            try {
+                sendCommand(new MPDCommand("noidle"));
+            } catch (IOException e) {
+            }
+        }
+    }
+
+    private void establishConnection() throws IOException, MPDException {
+        openSocket();
+
+        MPDCommand currentCommand = new MPDCommand("connect");
+        MPDResponse response = readResponse(currentCommand);
+
+        if (!response.isOk()) {
+            throw new MPDException("could not connect");
+        }
+
+        if (!password.isEmpty()) {
+            currentCommand = new MPDCommand("password", password);
+            sendCommand(currentCommand);
+            response = readResponse(currentCommand);
+            if (!response.isOk()) {
+                throw new MPDException("could not authenticate");
+            }
+        }
+    }
+
+    private void openSocket() throws IOException {
+        logger.debug("opening connection to {} port {}", this.address, this.port);
+        Socket socket = new Socket(this.address, this.port);
+
+        reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
```

Comment:
If I see correctly, you need to store `InputStreamReader` to close it, too.

Reply:
From the documentation it is not clear, whether the InputStreamReader is automatically closed by the BufferedReader. Looking at the code, it is closed. Since it does not harm, I have added it.